### PR TITLE
[Core][PCP][Output] Restore only sampled final rows after PCP prefill

### PR DESCRIPTION
**Pull request by foraxe**
## Summary

Restore only the final hidden-state rows needed for ordinary PCP sampling, instead of communicating every scheduled row. The fast multicast path now lives directly in the existing PCP manager: **there is no separate `PCPMulticastHiddenStateRestorer` class or transport module.**

```text
ASIS
rank-local hidden states -> AllGather all rows -> reorder all rows
  -> select final rows -> logits and sampling
```

```text
PR
existing PCP layout -> identify and pack final rows
  -> multicast + synchronization -> request-order rows -> logits and sampling
```

Pure decode uses already-replicated rows. Prompt logprobs, TP-sharded sampling and speculation retain dense restoration. Without multicast support, use the existing collective on compact rows; the one-row prefill case retains the ordinary collective. Persistent KV placement and quantization are unchanged; this is independent of #52863.

## Manager-owned lifecycle, same transport

This incorporates Lucas Wilkinson's structural cleanup from foraxe/vllm#5 while retaining multicast. The PCP manager is constructed before memory profiling and is the sole owner of its restore workspace. The model runner no longer allocates, stores or forwards restore buffers.

- Use the existing PCP manager for selected-row metadata, multicast buffers and output parity; no new class or module.
- Keep model-runner integration small and share prompt-logprob eligibility logic.
- Allocate packed/gathered/alternating-output buffers in the manager before KV profiling. Coordinate allocation and rendezvous failures across ranks; the manager synchronizes and releases its buffers on shutdown.
- Bind KV-dependent block tables later, updating the negotiated interleave and clearing stale batch metadata without recreating the manager or restore workspace. Repeated profiling/final initialization retains the same buffers.
- Skip multicast allocation when speculation or configured batch-sharded sampling requires the dense path.
- Preserve the three GPU operations: packing, PyTorch `multimem_all_gather_out`, and final reordering.
- Use existing token/request counts for the ordinary decode identity check. Retain lazy dense-index GPU upload so compact sampling does not pay for a full index upload.

Removing the class and removing multicast are separate decisions. The unchanged AllGather-only proposal matched correctness but showed +2.88% TTFT and >5% component slowdown in 22/24 cases in our bounded comparison, so this update retains multicast without its former abstraction.

## Controlled performance check

Reference: previous #49756 at `5e82de5fc`. Class-free candidate: `f935a4a9b`. The subsequent main merge and manager-ownership refactor retain the same multicast algorithm. Lifecycle/correctness checks were rerun; the historical performance numbers below are not a new benchmark of the ownership change.

Full GLM-5.2 NVFP4, four GB200 GPUs, Torch 2.13+cu130, eager MRV2, TP1PCP4EP4DCP1, FP8 KV, FlashInfer CUTLASS MoE, FLASH_ATTN prefill and unchanged sparse decode. Four requests of 1024 input/32 greedy output tokens, prefix cache disabled, chunked prefill enabled, 1024 KV blocks, model/batch limit 4096.

Four warmup batches plus six measured pairs in alternating order. Both implementations share model weights, KV, input and multicast buffers; metadata work remains part of the implementation comparison. Batches are fully queued before scheduling. All paired output tokens match, and every rank executes one prefill plus 31 decode steps per batch.

| Median | Previous PR | Class-free inline | Change |
| --- | ---: | ---: | ---: |
| Output throughput | 17.9516 tok/s | 17.8666 tok/s | -0.47% |
| TTFT | 276.778 ms | 278.110 ms | +0.48% |
| TPOT | 221.257 ms | 222.085 ms | +0.37% |

These results are within 0.5% on this workload, not a speedup or a statistically powered equivalence guarantee. They compare against the previous PR, not main; do not add them to earlier speedup figures.

Component coverage uses selected rows 1/16/64/256/1024/4096, BF16/FP16, balanced and single-owner-skewed layouts, eight warmup pairs and 30 measured pairs with the same multicast allocations. Outputs match. **It is not a uniform component win:** five of 24 median points are >5% slower, maximum +7.45%; minimum -6.38%. BF16 balanced 4096 is 187.38/187.58us; skewed 4096 is 413.68/400.21us. Small-row overhead remains visible rather than being hidden behind the engine result.

## Tests and evidence

```bash
CUDA_VISIBLE_DEVICES='' .venv/bin/python -m pytest tests/v1/worker/test_gpu_pcp_manager.py -q
# 96 passed, 2 GPU-only skipped
.venv/bin/python -m pytest tests/v1/worker/test_gpu_pcp_manager.py -q -k multicast_restore_buffers
# 1 passed: four GPUs, BF16/FP16, dense oracle, delayed rank,
# changing inputs, previous-output lifetime and repeated release
```

Lifecycle tests cover early allocation, workspace preservation across repeated KV binding, cleanup after failed binding, updated interleave, stale-state reset, always-dense configurations, and idempotent teardown. Rank-coordinated allocation/rendezvous failure coverage and scoped pre-commit checks, including mypy, pass.

A full GLM-5.2 NVFP4 startup/generation smoke test also passes on four GB200 GPUs with eager MRV2, TP1PCP4, 64 input tokens and four output tokens. It confirms the runner holds no separate buffer state, the manager is bound to the runner's KV tables and remains the same object, multicast executes, and output tokens exactly match compact collective restoration. This is lifecycle/correctness validation, not a new performance measurement.

Engine CUDA graphs, wider serving workloads and universal non-regression remain unvalidated. [Scripts, settings and raw results](https://github.com/foraxe/vllm/tree/agent/pcp-restore-lean-evidence/benchmarks/pcp_restore_inline) are kept outside the feature diff.

Lucas's commits and authorship are preserved. Implementation and validation used OpenAI Codex assistance. Co-authored with [@taoyuanyuan](https://github.com/taoyuanyuan).

### CODE REVIEW

**Comment by foraxe** [comment]
Updated this PR with the optimized multicast implementation ported to main at `9521c60bd`, retaining owned-row packing, reusable metadata/output buffers, lazy dense restoration, and the pure-decode bypass. Current head: `f36430039`; the final commit adds the engine reproduction harness.

The full GLM-5.2 NVFP4 comparison on 4x GB200 measured 651.98 -> 635.59 ms TTFT (-2.51%) and 17.84 -> 17.48 output tokens/s (-2.01%). Outputs matched exactly. Both sides of the tradeoff and the bounded measurement settings are in the description; this is not an overall serving-speedup claim.

I also checked #54030. Its separate direct-KV publish/wait launches are not present here: PyTorch's `multimem_all_gather_out` already performs multicast stores and cross-rank synchronization inside one CUDA kernel ([implementation at the tested Torch revision](https://github.com/pytorch/pytorch/blob/cf30153c4c131c8164ee7798e5022d810682e2cb/torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemoryOps.cu#L366)). Fusing selected-row packing or output reordering with multicast would be a distinct follow-up.

**Comment by foraxe** [comment]
Investigated the previously reported 2.01% throughput reduction. The original harness used immediate `LLM.generate()` submission, which could begin scheduling before all four prompts were queued. Instrumentation found different prefill/decode step counts between otherwise identical API batches.

I corrected the harness to pause scheduling, enqueue the entire batch, and resume. It now records TPOT separately; `--batch-submission immediate` reproduces the old behavior.

In a same-engine diagnostic with six alternating pairs, every measured batch executed one prefill plus 31 decode steps, and all output IDs matched. Main/multicast median throughput was 18.9459/19.0625 tok/s; median TPOT was 209.653/208.405 ms. Throughput favored multicast in five of six pairs (paired geometric change +1.27%). Decode restore CPU time dropped from approximately 178 us to 13 us per call.

The 2.01% loss did not reproduce under controlled scheduling. TTFT was slightly higher in this diagnostic (255.701 -> 258.924 ms), so the earlier TTFT gain should also remain qualified. This test keeps the multicast allocation resident in both modes and isolates execution of the algorithms, rather than their cold-start memory footprints. Production code is unchanged; commit `defd2cbee` fixes the benchmark submission method. The description retains both the initial results and these limitations.

**Review by LucasWilkinson — COMMENTED** [review]
@foraxe thanks for the PR I think we might be able to simplify this even further, some ideas: https://github.com/foraxe/vllm/pull/5

im not convinced we need the separate `PCPMulticastHiddenStateRestorer` abstraction

**Comment by foraxe** [comment]
Thanks @LucasWilkinson, the simplification is merged and tested.

File: vllm/v1/worker/gpu/model_runner.py [hunk 1]
```diff
@@ -291,6 +291,9 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
         self.draft_tokens_handler = DraftTokensHandler(self.device)
 
         self.pcp_manager: pcp.PCPManager | None = None
+        self.pcp_restore_buffers = pcp.allocate_pcp_restore_buffers(
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu/model_runner.py]: nit: can we please move this into the manager?

Comment (foraxe) [reply vllm/v1/worker/gpu/model_runner.py]: Moved in c7998a7f7, thanks! The PCP manager is now created before memory profiling and owns allocation, storage, and shutdown. KV initialization only binds/rebinds metadata without replacing the manager or its restore workspace; the runner no longer stores or forwards buffers. Always-dense configurations skip multicast allocation. The restore hot paths are unchanged. Validation: 96 CPU tests, four-GPU dense-oracle coverage, lint/mypy, and a full GLM-5.2 startup/generation smoke test with matching multicast/collective tokens.
